pkg/utils/gardener/project.go

```diff
@@ -46,15 +47,37 @@ func ProjectForNamespaceFromReader(ctx context.Context, reader client.Reader, na
 }
 
 // ProjectAndNamespaceFromReader returns the Project responsible for a given <namespace>. It reads the namespace and
-// fetches the project name label. Then it will read the project with the respective name.
+// fetches the project name label. Then it will read the project with the respective name or returns error if there is no
+// project for the given namespace.
 func ProjectAndNamespaceFromReader(ctx context.Context, reader client.Reader, namespaceName string) (*gardencorev1beta1.Project, *corev1.Namespace, error) {
 	namespace := &corev1.Namespace{}
 	if err := reader.Get(ctx, kutil.Key(namespaceName), namespace); err != nil {
 		return nil, nil, err
 	}
 
-	projectName := namespace.Labels[v1beta1constants.ProjectName]
-	if projectName == "" {
+	projectName, ok := namespace.Labels[v1beta1constants.ProjectName]
+	if !ok {
+		return nil, namespace, fmt.Errorf("project namespace doesn't have required label")
+	}
+
+	project := &gardencorev1beta1.Project{}
+	if err := reader.Get(ctx, kutil.Key(projectName), project); err != nil {
+		return nil, namespace, err
+	}
+
+	return project, namespace, nil
+}
+
+// NamespaceBelongsToAnyProject returns the Project responsible for a given <namespace>. It reads the namespace and
+// fetches the project name label. Then it will read the project with the respective name.
+func NamespaceBelongsToAnyProject(ctx context.Context, reader client.Reader, namespaceName string) (*gardencorev1beta1.Project, *corev1.Namespace, error) {
```

Comment:
What is your reasoning for duplicating the `ProjectAndNamespaceFromReader` function with only so little difference?

Reply:
In the admission plugin for releasing the namespace, we don't need the error if the project is not found because it checks if the given namespace doesn't have a required label for the project then it can be released.

Reply:
https://github.com/gardener/gardener/blob/ac97d589d5f5191dd1472ff70e3c1d0ccbd8947a/pkg/admissioncontroller/webhooks/admission/namespacedeletion/admission.go#L111-L124 I was not looking for better alternative but didn't find one.